crates/bevy_ecs/src/schedule/ambiguity_detection.rs

```diff
@@ -138,3 +279,223 @@ fn find_ambiguities(systems: &[impl SystemContainer]) -> Vec<(usize, usize, Vec<
     }
     ambiguities
 }
+
+#[cfg(test)]
+mod tests {
+    // Required to make the derive macro behave
+    use crate as bevy_ecs;
+    use crate::event::Events;
+    use crate::prelude::*;
+
+    #[derive(Resource)]
+    struct R;
+
+    #[derive(Component)]
+    struct A;
+
+    #[derive(Component)]
+    struct B;
+
+    // An event type
+    struct E;
+
+    fn empty_system() {}
+    fn res_system(_res: Res<R>) {}
+    fn resmut_system(_res: ResMut<R>) {}
+    fn nonsend_system(_ns: NonSend<R>) {}
+    fn nonsendmut_system(_ns: NonSendMut<R>) {}
+    fn read_component_system(_query: Query<&A>) {}
+    fn write_component_system(_query: Query<&mut A>) {}
+    fn with_filtered_component_system(_query: Query<&mut A, With<B>>) {}
+    fn without_filtered_component_system(_query: Query<&mut A, Without<B>>) {}
+    fn event_reader_system(_reader: EventReader<E>) {}
+    fn event_writer_system(_writer: EventWriter<E>) {}
+    fn event_resource_system(_events: ResMut<Events<E>>) {}
+    fn read_world_system(_world: &World) {}
+    fn write_world_system(_world: &mut World) {}
+
+    // Tests for conflict detection
+
+    #[test]
+    fn one_of_everything() {
+        let mut world = World::new();
+        world.insert_resource(R);
+        world.spawn().insert(A);
+        world.init_resource::<Events<E>>();
+
+        let mut test_stage = SystemStage::parallel();
+        test_stage
+            .add_system(empty_system)
+            // nonsendmut system deliberately conflicts with resmut system
+            .add_system(resmut_system)
+            .add_system(write_component_system)
+            .add_system(event_writer_system);
+
+        test_stage.run(&mut world);
+
+        assert_eq!(test_stage.ambiguity_count(&world), 0);
+    }
+
+    #[test]
+    fn read_only() {
+        let mut world = World::new();
+        world.insert_resource(R);
+        world.spawn().insert(A);
+        world.init_resource::<Events<E>>();
+
+        let mut test_stage = SystemStage::parallel();
+        test_stage
+            .add_system(empty_system)
+            .add_system(empty_system)
+            .add_system(res_system)
+            .add_system(res_system)
+            .add_system(nonsend_system)
+            .add_system(nonsend_system)
+            .add_system(read_component_system)
+            .add_system(read_component_system)
+            .add_system(event_reader_system)
+            .add_system(event_reader_system)
+            .add_system(read_world_system)
+            .add_system(read_world_system);
+
+        test_stage.run(&mut world);
+
+        assert_eq!(test_stage.ambiguity_count(&world), 0);
+    }
+
+    #[test]
+    fn read_world() {
+        let mut world = World::new();
+        world.insert_resource(R);
+        world.spawn().insert(A);
+        world.init_resource::<Events<E>>();
+
+        let mut test_stage = SystemStage::parallel();
+        test_stage
+            .add_system(empty_system)
```

Comment:
At first, while trying to understand each test and its expected output, I wondered why this system was added for this test.
(and I cloned your repository to run the test with report_ambiguities, to be sure I was not misunderstanding something with the expected value of 3).
So, personally, I would not have added `empty_system` here, or added a comment to say it should not conflict.
But is also fine as is, of course.

Reply:
I didn't write these tests, I just made them compile haha. I'm not entirely sure what the motivation was for adding the empty system, but it may have been to make sure it doesn't flag a false positive somehow. This would probably make more sense as its own test, though.

Reply:
Yep, that was the intent. It should be split out into it's own test though.

Reply:
Actually, I don't think it even needs its own test. This case is already covered in `read_only`, so I'll just remove the empty system from `read_world`.